File: quintus3.5/generic/qplib3.5/library/environ.c
/*  File   : environ.c
    Author : Richard A. O'Keefe
    Updated: 3/1/94
    Purpose: Provide access to UNIX environment variables,
	     and to the command line arguments.

    Copyright (C) 1987, Quintus Computer Systems, Inc.  All rights reserved.

    The function 'QAnvrn' is meant to be called only from
    environ.pl.  It backtracks over the elements of the
    "environ" vector, returning them as pairs of strings.
    This is rather hacky, so don't interrupt it!  After it
    is all over, the environ vector (hence getenv()) can
    be used again by C code.

    For access to the argument vector, we depend on the undocumented
    fact that exec builds the following structure in high memory:

		+--------+
		|  argc	 |	a 32-bit integer
		+--------+
		| argv[0]|	points to a string
		+--------+
		   ...
		+--------+
		| argv[n]|	last argument pointer, argc=n+1
		+--------+
		|  NULL  |	terminates argv[] sequence
		+--------+
*environ ------>| envv[0]|	first environment VAR=val string
		+--------+
		   ...
		+--------+
		| envv[m]|	m is not directly available
		+--------+
		|  NULL  |	terminates environment variables
		+--------+
		  bodies	The string bodies are higher

    That is, working backwards from environ, we find first NULL,
    then the arguments, which are pointers pointing HIGHER than
    environ, and finally the argument count, which is a smallish
    number.  In fact the maximum number of bytes in the argument
    list is 10240, and it takes at least 1 byte per argument,
    so this number must be in the range 0..10240.

    The variable argc is initialised to -1.  If it has some other
    value, it and the other variables have been initialised by
    QAargc.  We return -1 if anything looks suspicious.
*/

#include "quintus.h"
#include <ctype.h>
#include <stdlib.h>
#include <stdio.h>              /* [PM] 3.5 sprintf */
#include <string.h>             /* [PM] 3.5 strlen */

#ifndef	lint
static	char SCCSid[] = "@(#)94/03/01 environ.c	71.2";
#endif/*lint*/

#ifdef WIN32
#include <process.h>
#else
#include <unistd.h>             /* [PM] 3.5 getpid */
extern	char**	environ;
#endif

static	int	argc = -1;
static	char**	argv;	/* will point to argv[0] */
static	char**	argp;	/* runs along argv[] */


char *QAnext()
    {
	return *argp++;
    }


int QAargc()
    {
	if (argc < 0) {
#ifdef WIN32
	    argc = QP_argc;
	    argv = QP_argv;
#else /*unix*/
	    int k;
	    char *arg;

	    argv = environ;
	    if (*--argv) return -1;	/* should be NULL */
	    for (k = 0; ; k++) {
		if (k >= 10240) return -1;
		arg = *--argv;
		if (arg < (char*)environ) break;
	    }
	    if ((int)arg != k) return -1;
	    argc = k;
	    argv++;
#endif
	}
	argp = argv;
	return argc;
    }


/*  QA bige -- By Integer Get from Environment
    is given an index into environ and returns two strings and a result code.
    I rely on the index parameter being counted up from 0.
    Temporary assumption: no environment variable has more than 80
    characters in its name.
*/
int QAbige(index, variable, value)
    int index;
    char **variable;
    char **value;
    {
	static char buffer[81];
	register char *p, *q;
	
	if (!(p = environ[index])) {
	    *variable = "", *value = "";
	    return -1;
	}
	for (q = buffer; *p != '=' && (*q++ = *p++); ) ;
	if (!*p) {
	    *variable = "NO", *value = "=";
	    return -2;
	}
	*q = 0, p++;
	*variable = buffer, *value = p;
	return 0;
    }


/*  QA bnge  -- By Name Get from Environment
    is given a name to look up in the environment and returns a
    string and an error code.
*/
int QAbnge(variable, value)
    char *variable;
    char **value;
    {
	register char *p = getenv(variable);

	if (p) {
	    *value = p;
	    return 0;
	} else {
	    *value = "";
	    return -1;
	}
    }


/*  QA evxn  -- Environment Variable Expand Name
    is given a C string possibly containing references to environment
    variables, and returns a Prolog atom in which all such references
    have been expanded.  The possible cases are

	$<layout>	A dollar sign followed by a space, tab, or newline
			is taken as a literal dollar.  This is copied from
			the C shell.

	$$		A pair of dollar signs is replaced by the process
			identifier (see man 2 getpid) of the Prolog process
			expressed as 1-6 decimal digits.  This is copied
			from the Bourne and C shells both.

	$<alnums>	A sequence of letters, digits, and underscores
			is looked up in the environment, and replaced
			by the value that getenv() finds.

	$(stuff)	These four forms are identical in effect.  They
	$[stuff]	let you use environment variables with arbitrary
	$<stuff>	names, but their main point is that you can put
	${stuff}	a letter, digit, or underscore after them.

	$<other>	The dollar sign is taken literally, and the
			<other> character is processed.  (<layout>
			characters, you will recall, are discarded.)

    For example, suppose
	$HOST		is goonshow
	$USER		is bloodnok
	$HOME		is /usr/bloodnok
	$TERM		is vt100
    Then
	$HOME/thingy	is /usr/bloodnok/thingy
	${HOST}_${USER}	is goonshow_bloodnok
	$HOST_$USER	would look for a variable HOST_
	${TERM}-np	is vt100-np
	$<TERM>2	is vt1002
	$TERM2		would look for a variable TERM2
	/tmp/$USER$$	might be /tmp/bloodnok1234
	$ USER$		would be $USER$

    If all goes well, QAevxn returns 0.
    If an error is detected, it returns a character which identifies
    the error:
	')', ']', '>', or '}'	$( was found but the ) was missing, &c.
	'e'			An environment variable didn't exist.
	'o'			The result was too long for an atom.
*/
int QAevxn(source, result)
    char *source;
    QP_atom *result;
    {
	char buffer[QP_MAX_ATOM+8];
	char *limit = buffer+QP_MAX_ATOM;
	register char *s, *d, *t;
	register int c, x;

        /* initialize output argument to avoid a range error */
        *result = QP_atom_from_string("[]");

	s = source, d = buffer, t = limit;
	while ((c = *s++)) {
	    if (c != '$') {
		if (d >= t) return 'o';
		*d++ = c;
	    } else
	    switch (c = *s++) {
		default:
		    if (!isalnum(c)) {
			if (d >= t) return 'o';
			--s;
			*d++ = '$';
			break;
		    }
		    t = d;
		    do *d++ = c; while ((c = *s++), isalnum(c) || c == '_');
		    --s;
		    goto lookup;
		case ' ': case '\t': case '\n':
		    if (d >= t) return 'o';
		    *d++ = '$';
		    break;
		case '$':		/* $$ = process id */
		    x = getpid();
		    (void)sprintf(d, "%d", x);
		    d += strlen(d);
		    if (d >= t) return 'o';
		    break;
		case '(':
		    x = ')'; goto bracketed;
		case '[':
		    x = ']'; goto bracketed;
		case '<':
		    x = '>'; goto bracketed;
		case '{':
		    x = '}';
bracketed:	    t = d;
		    while ((c = *s++) && c != x) *d++ = c;
		    if (!c) return x;
lookup:		    *d = '\0';
		    d = getenv(t);
		    if (!d) return 'e';
		    if (t+strlen(d) >= limit) return 'o';
		    while ((c = *d++)) *t++ = c;
		    d = t, t = limit;
	    }
	}
	*d = '\0';
	*result = QP_atom_from_string(buffer);
	return 0;
    }

File: quintus3.5/generic/qplib3.5/library/environ.pl
%   Package: environ
%   Author : Richard A. O'Keefe
%   Updated: 31 Jan 1994
%   Purpose: provide access to UNIX environment variables
%	     and to the command line (argv).
%   SeeAlso: environ.c, man environ(5), man execve(2)

%   Copyright (C) 1987, Quintus Computer Systems, Inc.  All rights reserved.

/*  This module defines three predicates:

  (1)	environ(Name, Value)

    associates UNIX environment variables (Name) with their values
    (Value).  Both arguments are represented by Prolog atoms, NOT
    by chars.  environ/2 is a static predicate which can be used
    to look up a specified environment variable or to enumerate all
    the environment variables and their values.  Prior to 95.1 this
    was a dynamic table which was constructed at load time (so had    
    a static effect), in 95.1 and later it is a static predicate
    which checks the current environment (having a dynamic effect).

  (2)	argv(ArgumentList)

    unifies ArgumentList with a list of the arguments on the command
    line (see sh(1) or csh(1)) or from the exec call (see execve(2)).
    These arguments are represented by Prolog atoms.  In this release,
    you get the whole thing.  The first two arguments are always the
    absolute file name of the Prolog engine and the name of the saved
    state which is being run; remaining arguments are copied from the
    command line.

  (3)	expanded_file_name(NameWithVars, ExpandedVars)

    NameWithVars is an atom containing references to UNIX environment
    variables.  ExpandedVars is an atom obtained by substituting the
    appropriate values for the environment variables in NameWithVars.
    This is modelled on, but not identical to, variable substitution
    in the Bourne and C shells.  The possible cases are

	$<layout>	A dollar sign followed by a space, tab, or newline
			is taken as a literal dollar.  This is copied from
			the C shell.

	$$		A pair of dollar signs is replaced by the process
			identifier (see man 2 getpid) of the Prolog process
			expressed as 1-6 decimal digits.  This is copied
			from the Bourne and C shells both.

	$<alnums>	A sequence of letters, digits, and underscores
			is looked up in the environment, and replaced
			by the value that getenv() finds.

	$(stuff)	These four forms are identical in effect.  They
	$[stuff]	let you use environment variables with arbitrary
	$<stuff>	names, but their main point is that you can put
	${stuff}	a letter, digit, or underscore after them.

	$<other>	The dollar sign is taken literally, and the
			<other> character is processed.  (<layout>
			characters, you will recall, are discarded.)

    For example, suppose
	$HOST		is goonshow
	$USER		is bloodnok
	$HOME		is /usr/bloodnok
	$TERM		is vt100
    Then
	$HOME/thingy	is /usr/bloodnok/thingy
	${HOST}_${USER}	is goonshow_bloodnok
	$HOST_$USER	would look for a variable HOST_
	${TERM}-np	is vt100-np
	$<TERM>2	is vt1002
	$TERM2		would look for a variable TERM2
	/tmp/$USER$$	might be /tmp/bloodnok1234
	$ USER$		would be $USER$

    Bearing in mind that the built-in predicates such as open/3 and
    consult/1 do not understand environment variables (and _must_ not,
    because '$HOME' is a perfectly good UNIX file name as it stands),
    you will realise that the way to use expanded_file_name/2 is

	:- expanded_file_name('$HOME/$TERM.pl', File),
	   consult(File).

    At some future time these predicates might be made available as
	unix(environ(Name,Value))
    and unix(argv(ArgList)) respectively.  Please let us know if they
    would be useful enough to support that way.

    BEWARE: expanded_file_name/2 sees ONLY the environment variables.
    (These are the variables which have been 'export'ed by a Bourne
    shell, or set with 'setenv' in a C shell.  If environ/2 doesn't
    know about a variable, neither does expanded_file_name/2.)

    BEWARE: environ/2 should work on any UNIX system, but argv/1 has
    only been tested on the Sun 3.0 operating system.  It stands a
    good chance of working on other 4.2BSD-derived systems, but may
    not work on AIX or System V.
*/

:- module(environ, [
	argv/1,
	environ/2,
	expanded_file_name/2
   ]).
:- use_module(library(types), [
	must_be_symbol/3
   ]).

sccs_id('"@(#)94/01/31 environ.pl	71.1"').


%   argv(?Argv)
%   is true when Argv is a list of atoms representing *all* the arguments
%   which appeared in the UNIX command line.

argv(Argv) :-
	argc(Argc),
	Argc >= 0,
	pick_up_args(Argc, Argv).


pick_up_args(0, Argv) :- !,
	Argv = [].
pick_up_args(N, [Arg|Args]) :-
	M is N-1,
	pick_up_arg(Arg),
	pick_up_args(M, Args).



%   environ(?EnvVar:atom, ?ItsValue:atom)
%   is true when EnvVar is an atom naming a variable in the UNIX
%   environment (man 3 getenv) and ItsValue is an atom representing
%   the value of that environment variable.

environ(EnvVar, Value) :-
	(   var(EnvVar) ->
	    environ(0, EnvVar, Value)
	;   atom(EnvVar) ->
	    getenv(EnvVar, Value, 0)
	;   must_be_symbol(EnvVar, 1, environ(EnvVar,Value))
	).

environ(I, EnvVar, Value) :-
	getenv(I, Var, Val, 0),
	(   Value = Val, EnvVar = Var
	;   J is I+1,
	    environ(J, EnvVar, Value)
	).


%   expanded_file_name(+NameWithVars:atom, ?ExpandedName:atom)
%   replaces $var, $$, ${var} and so on in NameWithVars, yielding
%   ExpandedName.  E.g. expanded_file_name('$HOME/mytmp/QT$$a', TempName).

expanded_file_name(NameWithVars, ExpandedName) :-
	Goal = expanded_file_name(NameWithVars,ExpandedName),
	(   atom(NameWithVars) ->
	    dollar(NameWithVars, X, E),
	    (   E =:= 0 -> ExpandedName = X
	    ;   raise_exception(
		    domain_error(Goal,1,filename,NameWithVars,
			expanded_file_name(E)))
	    )
	;   must_be_symbol(NameWithVars, 1, Goal)
	).

:- multifile 'QU_messages':msg/3.

'QU_messages':msg(expanded_file_name(E)) --> 
        {expand_error(E, Message)},
	[Message-[]].

expand_error(0'),	'Missing ) after $(').
expand_error(0'],	'Missing ] after $[').
expand_error(0'>,	'Missing > after $<').
expand_error(0'},	'Missing } after ${').
expand_error(0'e,	'Unknown environment variable').
expand_error(0'o,	'Expansion too long').


foreign_file(library(system(libpl)), [
	'QAargc', 'QAnext', 'QAbnge', 'QAbige', 'QAevxn'
]).

foreign('QAargc', argc([-integer])).
foreign('QAnext', pick_up_arg([-string])).
foreign('QAbnge', getenv(+string,-string,[-integer])).
foreign('QAbige', getenv(+integer,-string,-string,[-integer])).
foreign('QAevxn', dollar(+string,-atom,[-integer])).

:-  load_foreign_executable(library(system(libpl))),
    abolish(foreign_file, 2),
    abolish(foreign, 2).

